### PR TITLE
Enable metered verifier at signing

### PR DESCRIPTION
**Pull request by oxade**
## Description 
Adds a metered verifier at signing. Verification only fails if there's a timeout.

Next: 
Will create a Move PR to improve the specificity of verifier timeout status code.
More tests

## Test Plan 

Added test to ensure limits hit.
More coming in subsequent PR.
---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [x] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-adapter/src/adapter.rs [hunk 1]
```diff
@@ -42,39 +44,64 @@ use sui_types::{
...
+    is_metered: bool,
+) -> VerifierConfig {
+    let (
+        max_back_edges_per_function,
+        max_back_edges_per_module,
+        max_per_fun_meter_units,
+        max_per_mod_meter_units,
+    ) = if is_metered {
+        // TODO: change these to pull from protocol config
+        (None, None, Some(1000 * 8000), Some(1000 * 8000))
```
Comment (kchalkias) [inline crates/sui-adapter/src/adapter.rs:66]: I understand we''l have to eventually load them from protocol config, but pending that PR, generally prefer setting any numeric static value to a const for readability.

Comment (oxade) [reply crates/sui-adapter/src/adapter.rs:66]: > I understand we''l have to eventually load them from protocol config, but pending that PR, generally prefer setting any numeric static value to a const for readability. 

Yup. Not to worry though, I will put the consts in protocol config before the PR is ready for review.

**Review by dariorussi — COMMENTED** [review]
great work!
A couple of comments, it would be nice to have a test that fails.
Waiting for others to comment as well and I'll come back later for approval

File: crates/sui-adapter/src/adapter.rs [hunk 2]
```diff
@@ -727,3 +758,55 @@ pub fn missing_unwrapped_msg(id: &ObjectID) -> String {
...
+            ]
+            .contains(&e.major_status())
+            {
+                return Err(SuiError::ModuleVerificationFailure {
+                    error: "Verification timedout".to_string(),
+                });
+            };
+        }
+    }
+    // Run Sui bytecode verifier, which runs some additional checks that assume the Move bytecode verifier has passed.
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-adapter/src/adapter.rs:805]: Ah good point. I meant to pose a question there of "do we care to meter this too right now"?

File: crates/sui-core/src/transaction_input_checker.rs [hunk 3]
```diff
@@ -55,6 +56,7 @@ pub async fn check_transaction_input(
 ) -> SuiResult<(SuiGasStatus<'static>, InputObjects)> {
     transaction.check_version_supported(epoch_store.protocol_config())?;
     transaction.validity_check(epoch_store.protocol_config())?;
+    check_non_system_packages_to_be_published(transaction, epoch_store.protocol_config())?;
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-core/src/transaction_input_checker.rs:59]: Good point.

[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-core/src/transaction_input_checker.rs:59]: Done

**Comment by oxade** [comment]
> great work! A couple of comments, it would be nice to have a test that fails. Waiting for others to comment as well and I'll come back later for approval

I've added a test now.

File: crates/sui-core/src/transaction_input_checker.rs [hunk 4]
```diff
@@ -384,3 +386,21 @@ fn check_one_object(
...
+#[instrument(level = "trace", skip_all)]
+pub fn check_non_system_packages_to_be_published(
+    transaction: &TransactionData,
+    protocol_config: &ProtocolConfig,
+) -> UserInputResult<()> {
+    // Only meter non-system TXes
+    if !transaction.is_system_tx() {
+        if let TransactionKind::ProgrammableTransaction(pt) = transaction.kind() {
+            pt.non_system_packages_to_be_published()
+                .try_for_each(|q| run_metered_move_bytecode_verifier(q, protocol_config))
```
Comment (tzakian) [inline crates/sui-core/src/transaction_input_checker.rs:400]: TIL about `try_for_each`!

**Review by tzakian — APPROVED** [review]
Overall looks good to me. 

Only thing I'd ask you take a look at/consider is the name of the function -- I found it pretty confusing looking at the name/return type figuring out what it was doing at first.

File: crates/sui-adapter/src/adapter.rs [hunk 5]
```diff
@@ -727,3 +755,56 @@ pub fn missing_unwrapped_msg(id: &ObjectID) -> String {
         id
     )
 }
+
+/// Run the bytecode verifier with a meter limit
+/// This function only fails if the verification does not complete within the limit
+#[instrument(level = "trace", skip_all)]
+pub fn run_metered_move_bytecode_verifier(
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-adapter/src/adapter.rs:762]: Good callouts. If its okay, I'll add make these changes in a fast follow PR, so we can at least get this in.

File: crates/sui-adapter/src/adapter.rs [hunk 6]
```diff
@@ -727,3 +755,56 @@ pub fn missing_unwrapped_msg(id: &ObjectID) -> String {
...
+            // TODO: currently the Move verifier emits `CONSTRAINT_NOT_SATISFIED` for various failures including metering timeout
+            // We need to change the VM error code to be more specific when timedout for metering
+            if [
+                StatusCode::CONSTRAINT_NOT_SATISFIED,
+                StatusCode::TOO_MANY_BACK_EDGES,
+            ]
+            .contains(&e.major_status())
+            {
+                return Err(SuiError::ModuleVerificationFailure {
+                    error: "Verification timedout".to_string(),
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-adapter/src/adapter.rs:804]: Will address in follow up PR :)

**Pull request merged**